Savestates: Use LZ4 algorithm for faster decompression

malleoz opened the pull request:
Compared to LZO, LZ4 has near-identical compression ratio and compression time. However, LZ4 boasts substantial decompression time improvement. While compression time doesn't matter much for TASers, repeated savestate loading requires decompression to be fast, in order to improve efficiency during TAS sessions. While [their website](https://lz4.org/) shows benchmarks that produce a 67% speedup in decompression time, in practice I found that for savestates it produces typical speedups of about 72%. This improvement is larger when the Internal Resolution is turned up higher, as this increases the typical savestate filesize.

LZ4 can compress at most `LZ4_MAX_INPUT_SIZE` ~= 2.11GB at a time. To handle the upper-bound, I compress `LZ4_MAX_INPUT_SIZE` at a time, looping each time so long as we have more bytes to compress. This scenario is very unlikely to affect anyone, as I had to increase my Internal Resolution to about 20x for this to occur on my end.

Note that Netplay also uses LZO for packet transfer of the memcard between players. Given that I have no experience with this feature, and to prevent possible regression from limited testing due to my lack of experience, I've chosen to leave Netplay untouched. Note that if we change Netplay to use LZ4, then the LZO dependency can be removed from the project.

Support has been kept for forwards/backwards compatibility. These changes support reading the version info for legacy LZO savestates in order to show the user what version the incompatible state was made on. Loading new LZ4 states on older versions of Dolphin results in the version string still being shown to the user in an error message.

These changes aim to future-proof the ability the add more uncompressed data to the header. We can adjust the size of `StateExtendedHeader` by incrementing `StateExtendedBaseHeader.payload_offset`. Thus we can support the serialization of any new fields, including char arrays.

AdmiralCurtiss commented:
Please use a submodule for lz4 and break out adding it into its own commit.

AdmiralCurtiss commented:
(I guess I could be convinced to have lz4 as not-submodule if you reduce it down to just the source + license without any of the extra files, it's pretty small and doesn't update all that often. But definitely break it out to its own commit.)

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -388,32 +371,28 @@ static void CompressAndDumpState(CompressAndDumpState_args& save_args)
-      {
-        cur_len = IN_LEN;
-      }
+      u32 bytes_to_compress = std::min((u32)LZ4_MAX_INPUT_SIZE, header.size - bytes_compressed);
```
It's probably worth adding a utility 'compress buffer to file' function instead of having the compression logic inline like this. I know that's what the old code did but that code is ancient and we have different standards now.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -388,32 +371,28 @@ static void CompressAndDumpState(CompressAndDumpState_args& save_args)
-        cur_len = IN_LEN;
-      }
+      u32 bytes_to_compress = std::min((u32)LZ4_MAX_INPUT_SIZE, header.size - bytes_compressed);
+      char* out = new char[bytes_to_compress];
```
Also please no raw new/delete.

AdmiralCurtiss replied:
https://github.com/lz4/lz4/blob/dev/lib/lz4.h#L215-L224C7

malleoz replied:
I did see that bound function, but given that our size is originally an u32, if it exceeds the signed int limit it will fail the bounds check and it will output 0.

malleoz replied:
I guess if it outputs 0 then we know to set the limit to `LZ4_MAX_INPUT_SIZE`

malleoz commented:
> (I guess I could be convinced to have lz4 as not-submodule if you reduce it down to just the source + license without any of the extra files, it's pretty small and doesn't update all that often. But definitely break it out to its own commit.)

There's a lot of tinkering in regards to the CMakeLists.txt that I wouldn't say I'm comfortable safely trimming excess code from, so I will adjust this to be a submodule.

Pokechu22 commented:
> Though Dolphin stores the emulator version in the savestate, so as to disallow savestate loads across different versions, the version is stored in the COMPRESSED part of the savestate. Thus, changing the compression algorithm means that this will not be read before decompression is attempted.

Does this apply to the `STATE_VERSION` field?

https://github.com/dolphin-emu/dolphin/blob/da6339a72275f6b499a003b5486ebfc8f26e0d2f/Source/Core/Core/State.cpp#L98-L99

I'd hope that that's in an uncompressed header since that's the main thing that determines incompatibility (if the exact emulator version doesn't match it'll still try to load, but if the state version doesn't match it'll reject it). I'm not sure if it actually is or not though.

malleoz commented:
> > Though Dolphin stores the emulator version in the savestate, so as to disallow savestate loads across different versions, the version is stored in the COMPRESSED part of the savestate. Thus, changing the compression algorithm means that this will not be read before decompression is attempted.
> 
> Does this apply to the `STATE_VERSION` field?
> 
> https://github.com/dolphin-emu/dolphin/blob/da6339a72275f6b499a003b5486ebfc8f26e0d2f/Source/Core/Core/State.cpp#L98-L99
> 
> I'd hope that that's in an uncompressed header since that's the main thing that determines incompatibility (if the exact emulator version doesn't match it'll still try to load, but if the state version doesn't match it'll reject it). I'm not sure if it actually is or not though.

The check against `STATE_VERSION` is performed after decompression occurs sadly. Although I guess I might as well update the state version in this PR as well, for posterity's sake.

Pokechu22 commented:
Hmm. I'd prefer it if there was a backwards-compatible way of doing things (in that trying to load the savestate on an older version gives a suitable error message).

I think this could theoretically be done by either LZO-compressing only the version number or by disabling compression on the version number entirely (since a size of 0 in the header is treated as not compressed). I *think* we could have whatever data we want after it detects an incorrect version, but I'm not 100% sure. There are two reserved fields in the header, so it would be possible to use them for a new format identifier or something. I'm not entirely sure what that would look like and it would require testing against older versions to make sure things work; maybe the format this PR currently uses gives a good enough error message.

One other thing: Dolphin currently includes [Zstd](https://en.wikipedia.org/wiki/Zstd) since the rvz format uses it. This is supposed to be very fast for decompression; it might be better to use it instead of adding a new algorithm. If LZ4 is faster or about the same in practice though then an extra dependency is probably fine.

malleoz commented:
> I think this could theoretically be done by either LZO-compressing only the version number or by disabling compression on the version number entirely (since a size of 0 in the header is treated as not compressed).

If I understand correctly, lzo decompresses in chunks of 65536 bytes. To retrieve the version number you'd have to decompress that entire block, which seems wasteful.

> There are two reserved fields in the header, so it would be possible to use them for a new format identifier or something. I'm not entirely sure what that would look like and it would require testing against older versions to make sure things work; maybe the format this PR currently uses gives a good enough error message.

Storing the revision in the header would be a good idea, but there'd definitely need to be discussion on how we would define this.

> One other thing: Dolphin currently includes Zstd

They are in the same family of compression algorithms, but lz4 outperforms zstd in terms of decompression speeds. I don't have exact metrics, but here's one source: https://indico.fnal.gov/event/16264/contributions/36466/attachments/22610/28037/Zstd__LZ4.pdf

malleoz commented:
Rebased because I realized I had applied CMake-related changes in the second commit accidentally. This should be properly separated now.

JosJuice commented:
> I *think* we could have whatever data we want after it detects an incorrect version, but I'm not 100% sure.

That should be possible, but make sure it's after the version string, not just after the version number.

> If I understand correctly, lzo decompresses in chunks of 65536 bytes. To retrieve the version number you'd have to decompress that entire block, which seems wasteful.

Then you could disable compression. The part of the file that needs to be readable by old Dolphin versions is only 20 bytes or so.

AdmiralCurtiss commented on CMake/FindLZ4.cmake:
```diff
@@ -0,0 +1,15 @@
+find_path(LZ4_INCLUDE_DIR lz4.h)
```
Is this file needed at all? I would expect modern CMake to be able to find the system lz4 by itself, and our externals should be taken care of with the `dolphin_find_optional_system_library()` call.

malleoz replied:
Yes I can remove this and will do so on my next rebase.

AdmiralCurtiss replied:
Okay I tested this and yeah it's not needed. I also noticed a few other CMake issues, this should fix all of them: https://github.com/AdmiralCurtiss/dolphin/commit/1539fa378136b8987cf710f8e6187e4daeaa22ed

AdmiralCurtiss commented on Externals/lz4/LZ4.vcxproj:
```diff
@@ -0,0 +1,128 @@
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|Win32'">
+    <ClCompile>
+      <WarningLevel>Level4</WarningLevel>
+      <PrecompiledHeader>
+      </PrecompiledHeader>
+      <Optimization>MaxSpeed</Optimization>
+      <FunctionLevelLinking>true</FunctionLevelLinking>
+      <IntrinsicFunctions>true</IntrinsicFunctions>
+      <PreprocessorDefinitions>WIN32;NDEBUG;LZ4_DLL_EXPORT=1;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <TreatWarningAsError>false</TreatWarningAsError>
+      <EnablePREfast>false</EnablePREfast>
+      <RuntimeLibrary Condition="'$(UseStaticCRT)'=='true'">MultiThreaded</RuntimeLibrary>
+    </ClCompile>
+    <Link>
+      <GenerateDebugInformation>true</GenerateDebugInformation>
+      <EnableCOMDATFolding>true</EnableCOMDATFolding>
+      <OptimizeReferences>true</OptimizeReferences>
+    </Link>
+  </ItemDefinitionGroup>
+  <ItemDefinitionGroup Condition="'$(Configuration)|$(Platform)'=='Release|x64'">
+    <ClCompile>
+      <WarningLevel>Level4</WarningLevel>
+      <PrecompiledHeader>
+      </PrecompiledHeader>
+      <Optimization>MaxSpeed</Optimization>
+      <FunctionLevelLinking>true</FunctionLevelLinking>
+      <IntrinsicFunctions>true</IntrinsicFunctions>
+      <PreprocessorDefinitions>WIN32;NDEBUG;LZ4_DLL_EXPORT=1;%(PreprocessorDefinitions)</PreprocessorDefinitions>
+      <TreatWarningAsError>false</TreatWarningAsError>
+      <EnablePREfast>true</EnablePREfast>
+      <AdditionalOptions>/analyze:stacksize295252 %(AdditionalOptions)</AdditionalOptions>
+      <RuntimeLibrary Condition="'$(UseStaticCRT)'=='true'">MultiThreaded</RuntimeLibrary>
+    </ClCompile>
+    <Link>
+      <GenerateDebugInformation>true</GenerateDebugInformation>
+      <EnableCOMDATFolding>true</EnableCOMDATFolding>
+      <OptimizeReferences>true</OptimizeReferences>
+    </Link>
+  </ItemDefinitionGroup>
```
Is all of this really needed? Comparing with our other externals VS files I don't think it is.

AdmiralCurtiss commented:
In case you missed this in the conversation you resolved, your CMake definitions are a bit wrong in general and don't work correctly with a system-installed lz4, I've fixed this here: https://github.com/AdmiralCurtiss/dolphin/commit/1539fa378136b8987cf710f8e6187e4daeaa22ed

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -354,6 +337,25 @@ static std::map<double, int> GetSavedStates()
   return m;
 }
 
+static int CompressBufferToFile(const u8* raw_buffer, u32 max_bytes_to_compress, File::IOFile& f)
```
I was kinda hoping you'd include the entire loop in the helper function, so the user doesn't have to care about that and can just compress arbitrarily-sized data with a single call.

malleoz replied:
Sure I can do that no problem!

malleoz commented:
> In case you missed this in the conversation you resolved, your CMake definitions are a bit wrong in general and don't work correctly with a system-installed lz4, I've fixed this here: [AdmiralCurtiss@8f82639](https://github.com/AdmiralCurtiss/dolphin/commit/8f82639bcd8f77c952ea8bc635d9e84520499d27)

Thank you immensely for doing this. I'll apply these changes on my next rebase

malleoz commented:
Updated my commits as follows:
- Apply changes from @AdmiralCurtiss to fix CMake
- Tuck compression/decompression logic inside helper functions, who loop while we still have bytes to compress/decompress
- Migrate to using a different field in the header for size, leaving the previous field zero'd out
- Leave version info uncompressed
  - Because the old size field is now zero'd out, old Dolphin will treat these new savestates as uncompressed. Thus it will read in the entire file, including the version info. Loading a new savestate format on older Dolphin versions will detect version mismatch and bail out.

I guess the last thing to debate is if we want to reserve empty space after the version info for future changes... I believe we don't need to do this. Any new data occurring after the version info and before the compressed data would result in the following:
- lz4 tries to decompress starting at this new data
- the compression fails, so we bail out
- The buffer has been written to, so in `State::DoState`, the version check will occur
- The version check will fail, so the state will not load

AdmiralCurtiss commented:
I think that if we're already changing the format here we should actually do it right. I'm gonna need to think about a layout that makes sense and remains compatible with being read by older versions, but I think if we embed a state format version and an offset to the start of the 'payload' (ie, compressed part) somewhere in the uncompressed 'header' we should be pretty future-proof.

AdmiralCurtiss commented:
Oh and uh, you seem to have removed the ability for newer versions of Dolphin to decompress LZO compressed states, which is kinda needed if we want to display the version number to the user. It can definitely be removed eventually in the future, but I think for now it should stay in.

malleoz commented:
> Oh and uh, you seem to have removed the ability for newer versions of Dolphin to decompress LZO compressed states, which is kinda needed if we want to display the version number to the user. It can definitely be removed eventually in the future, but I think for now it should stay in.

Once you determine a suitable header layout, I can add this back in. I imagine it's as simple as checking to see if the old `header.size` field is non-zero, as this indicates it was set via LZO

AdmiralCurtiss commented:
I wrote this up yesterday and forgot to post it. On review I might have overdesigned this a little, use your own judgement I suppose.

---

Okay, I looked through the state code a bit and I think I have a pretty good idea what we can do now.

For reference:

Current Dolphin:
https://github.com/dolphin-emu/dolphin/blob/c86a0a04dd355ba9c7c475ae89c50cbff2e88d31/Source/Core/Core/State.cpp
https://github.com/dolphin-emu/dolphin/blob/c86a0a04dd355ba9c7c475ae89c50cbff2e88d31/Source/Core/Core/State.h

Oldest version that current Dolphin can still read:
https://github.com/dolphin-emu/dolphin/blob/ef2e0a87d06eb498f0e2eaebe6be2785740e54dd/Source/Core/Core/Src/State.cpp
https://github.com/dolphin-emu/dolphin/blob/ef2e0a87d06eb498f0e2eaebe6be2785740e54dd/Source/Core/Core/Src/State.h

So, first thing: Old versions of Dolphin ([older than ~3.5 years](https://github.com/dolphin-emu/dolphin/commit/7885fdb1ae7002f7c25a35bfc670b3f2b470d203)) [do not clear the reserved bytes](https://github.com/dolphin-emu/dolphin/blob/ef2e0a87d06eb498f0e2eaebe6be2785740e54dd/Source/Core/Core/Src/State.cpp#L246-L251) in the header, so we cannot use them as an indicator for the new version, as they may have been written as anything by older versions. This is a little bit annoying, but ultimately not a huge deal.

As mentioned before, the state version is in the compressed area, and [the LZO compressedness](https://github.com/dolphin-emu/dolphin/blob/c86a0a04dd355ba9c7c475ae89c50cbff2e88d31/Source/Core/Core/State.cpp#L597) [is determined by the size field](https://github.com/dolphin-emu/dolphin/blob/ef2e0a87d06eb498f0e2eaebe6be2785740e54dd/Source/Core/Core/Src/State.cpp#L371), so the size field has to be zero to remain compatible with the old reader. The old payload [starts with the version info](https://github.com/dolphin-emu/dolphin/blob/c86a0a04dd355ba9c7c475ae89c50cbff2e88d31/Source/Core/Core/State.cpp#L132) and [further reading is cancelled](https://github.com/dolphin-emu/dolphin/blob/ef2e0a87d06eb498f0e2eaebe6be2785740e54dd/Source/Core/Core/Src/State.cpp#L88-L95) [if that does not match](https://github.com/dolphin-emu/dolphin/blob/c86a0a04dd355ba9c7c475ae89c50cbff2e88d31/Source/Core/Core/State.cpp#L174-L183), so after that we can do whatever. So assuming an non-LZO compressed state, the header we have to match is:

```
char[6] gameID
byte[2] arbitrary
u32 must be zero
byte[4] arbitrary
double timestamp
u32 state version
u32 length of following string
char[^length] dolphin version that made the state
```

Our new 'extended header' should go after that. I propose something along the lines of:

```
u16 extended header revision
u16 compression type of payload (0 = uncompressed, 1 = Lz4, higher numbers reserved)
u32 offset of payload
u64 length of payload (uncompressed)
u64 length of payload (compressed -- should be same as the previous value if payload is uncompressed)
<--- offset is relative to this position. the fields above this are considered the 'base extended header'
byte[^offset] space for extra uncompressed data if needed
byte[^length] payload
```

Will we ever need to read the payload on a mismatching extended header revision or anything fancy like that? Who knows, but it's good to be prepared just in case -- maybe we will improve our savestate system someday so we can actually load states from different Dolphin versions to some degree!

I was also considering adding an extra timestamp after the compressed length that isn't in [this extremely stupid format](https://github.com/dolphin-emu/dolphin/blob/c86a0a04dd355ba9c7c475ae89c50cbff2e88d31/Source/Core/Core/State.cpp#L301-L328) but I'm not sure if that really matters. The current format is stupid, but it won't actually cause issues until ~280k years from now if my math is right...

---

Reading a state has to be done in a few steps in order to be able to read all formats going back to state version 17 correctly. Basically:

- Read 'old' 24-byte header.
  - If size field is nonzero, the rest of the data is LZO compressed. Decompress the first handful bytes (not sure how much control we have over this) in order to determine the Dolphin version string to display to the user, then abort the load regardless of what version is reported in there.
  - If size field is zero, continue.
- [Read state version.](https://github.com/dolphin-emu/dolphin/blob/c86a0a04dd355ba9c7c475ae89c50cbff2e88d31/Source/Core/Core/State.cpp#L134-L140)
- [Read or construct Dolphin version string.](https://github.com/dolphin-emu/dolphin/blob/c86a0a04dd355ba9c7c475ae89c50cbff2e88d31/Source/Core/Core/State.cpp#L142-L162)
- Check state version.
  - If it mismatches the current Dolphin's state version, abort the load and report to user.
  - If it matches, continue.
- Read the base extended header.
- Skip [offset of payload] bytes.
- Read [length of payload (compressed)] bytes into buffer.
  - If compression type != 0 and type is known, decompress buffer into second buffer then continue with that buffer.
  - If compression type != 0 and type is not known, abort load.
  - If compression type == 0 just continue with the buffer we just read.
- Finally, take buffer and hand off to PointerWrap/DoState() for actual savestate load.

malleoz commented:
> u64 length of payload (uncompressed)
> u64 length of payload (compressed -- should be same as the previous value if payload is uncompressed)

The only hold-up I have here is that if we exceed the maximum number of bytes that can be compressed into a single payload, then there will be "multiple" payloads. In this scenario, we don't necessarily know the length of each individual compressed payload solely based off the size of the uncompressed data. Maybe length of payloads can be stored as an array of u64, but then we'd also need a separate variable to know the number of payloads present.

malleoz commented:
Alrighty, I've appended the version information to the original header. Note that because the version string is a `char*`, some arithmetic on reading/writing the header is uglier than before. I've added the extended header, factoring in future possibility of adjusting the `payload_offset` for more uncompressed data. I've tested the following scenarios which all work fine:
- Loading super old savestates will access `s_old_versions` and correctly display the version range
- Loading savestates with a different version string will error, showing the user that version string
- Loading current savestates (compressed/uncompressed) succeed, including scenarios where we exceed `LZ4_MAX_INPUT_SIZE`
- Loading current savestates on older versions of Dolphin shows the version string

Note that, like I said in my previous comment, I do not include the compressed size in the extended header. Instead I continuously read bytes so long as we have not read `uncompressed_size` bytes.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -354,6 +306,58 @@ static std::map<double, int> GetSavedStates()
   return m;
 }
 
+static void CompressBufferToFile(const u8* raw_buffer, u32 size, File::IOFile& f)
+{
+  u32 total_bytes_compressed = 0;
```
`size`, `total_bytes_compressed`, `bytes_left_to_compress` should be `u64`, there's no reason to limit that to 4GB.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -354,6 +306,58 @@ static std::map<double, int> GetSavedStates()
+      PanicAlertFmtT("Internal LZ4 Error - compression failed");
+
+    // The size of the data to write is 'compressed_len'
+    f.WriteArray(&compressed_len, 1);
```
Although it's exceptionally unlikely that `int` is anything but 32 bits, it's probably still a good idea to ensure that since you're serializing it out.

malleoz replied:
I assume changing `compressed_len` to a `int32_t` suffices?

AdmiralCurtiss replied:
We usually use `s32` in Dolphin but yeah that's fine.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -525,13 +493,69 @@ void SaveAs(const std::string& filename, bool wait)
+  if (res != LZO_E_OK)
+  {
+    // This doesn't seem to happen anymore.
+    PanicAlertFmtT("Internal LZO Error - decompression failed ({0}) ({2}) \n"
+                   "Unable to retrieve outdated savestate version info.",
+                   res, new_len);
```
This fmt string is not legal, I'm surprised this wasn't caught by a compiletime assert somewhere. There is no third argument matching `{2}`.

malleoz replied:
I can't believe I missed that, good catch! This was left in after I modified the old LZO panic alert.

AdmiralCurtiss commented on Source/Core/Core/State.h:
```diff
@@ -20,15 +21,40 @@ static const u32 NUM_STATES = 10;
+  u32 must_be_zero = 0;  // prevent decompression so older Dolphin can read state version info
+  char reserved2[4];
   double time;
+  u32 version_cookie;
+  u32 version_str_len;
+  std::unique_ptr<char[]> version_str;
+
+  int FileLength() { return offsetof(StateHeader, version_str) + version_str_len; }
```
I suspect it is not legal in C++ to have a partially-copyable types like this. You should probably split this up into multiple nested structs.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -525,13 +493,69 @@ void SaveAs(const std::string& filename, bool wait)
+  }
+
+  // Read in cookie, string length, and the string
+  memcpy(&header.version_cookie, buffer.data(), 2 * sizeof(u32));
+  header.version_str = std::make_unique<char[]>(header.version_str_len);
+  memcpy(header.version_str.get(), buffer.data() + 2 * sizeof(u32), header.version_str_len);
```
This isn't bounds checked.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -525,13 +493,69 @@ void SaveAs(const std::string& filename, bool wait)
+  else
+  {
+    // Read in cookie, string length, and the string
+    f.ReadArray(&header.version_cookie, 2);
+    header.version_str = std::make_unique<char[]>(header.version_str_len);
+    f.ReadBytes(header.version_str.get(), header.version_str_len);
```
You should check for the success of the reads.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -525,13 +493,69 @@ void SaveAs(const std::string& filename, bool wait)
+
+  if (!f.ReadBytes(&header, cookie_offset))
+  {
+    Core::DisplayMessage("State not found", 2000);
```
There's probably some improvements to be made here regarding error reporting. 'no file with that name exists' is a different scenario than 'there is a file here but we don't recognize it as a valid savestate'.

malleoz commented:
Updated to include @AdmiralCurtiss latest suggestions.
- Allow compressing/decompressing beyond 4GB by using `u64` instead of `u32`
- Validate we are serializing a 4 byte int for the length of each payload
- Fixed illegal fmt expression
- Broke headers into sub-structs to more clearly be able to write each header and determine flattened data length
- Add bounds checking prior to doing some `memcpy` operations
- Add more checks for failing to read from file
- Tweak error messages to be a bit more helpful (differentiate between file not found and no more bytes to read)

malleoz commented:
Updated to remove `<memory>` import since we changed the structs to not store a `unique_ptr` anymore

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -354,6 +306,71 @@ static std::map<double, int> GetSavedStates()
+
+  f.WriteArray(&extended_header.base_header, 1);
+  // If StateExtendedHeader is ammended to include more than the base, add WriteBytes() calls here.
+  // Do NOT use WriteArray, as this would prevent future revisions from writing char
+  // arrays stored in the extended header.
```
I don't understand what you're saying here.

malleoz replied:
So the old code wrote the header to the file via `f.WriteArray(&header, 1)`. This is essentially a memcpy. What I'm trying to say is that, in order to support a future extended header containing a `std::string`, we can't just memcpy the whole thing. Instead we should use `f.WriteBytes()` for every field in the header. This might be an unnecessary comment though.

JosJuice replied:
Yes, I think that comment is unnecessary. What function we use for writing the header is something that can be changed in the future anyway.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -354,6 +306,71 @@ static std::map<double, int> GetSavedStates()
+  f.WriteString(header.version_string);
+
+  f.WriteArray(&extended_header.base_header, 1);
+  // If StateExtendedHeader is ammended to include more than the base, add WriteBytes() calls here.
```
It's just one `m` in `amended`.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -525,13 +506,107 @@ void SaveAs(const std::string& filename, bool wait)
+  lzo_uint32 cur_len = 0;  // size of compressed bytes
+  lzo_uint new_len = 0;    // size of uncompressed bytes
+  std::vector<u8> buffer;
+  buffer.resize(header.legacy_header.must_be_zero);
```
This looks really weird... maybe rename `must_be_zero` to `lzo_size` or something like that instead, but keep a comment about it needing to be zero for newly created states.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -525,13 +506,107 @@ void SaveAs(const std::string& filename, bool wait)
+{
+  if (!f.IsOpen())
+  {
+    Core::DisplayMessage("State not found", 2000);
```
This is missing a `return false;`

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +629,99 @@ u64 GetUnixTimeOfSlot(int slot)
+    auto compressed_data = std::make_unique<char[]>(compressed_data_len);
+    f.ReadBytes(compressed_data.get(), compressed_data_len);
+
+    // We need to specify the output buffer's size for safety. This may exceed the positive bound
+    // of int, causing the buffer size to be interpreted as a negative value.
```
I don't know what this comment has to do with anything.

malleoz replied:
Yeah this is a bit unnecessary. I think it's clear what the operation below it is doing.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +629,99 @@ u64 GetUnixTimeOfSlot(int slot)
+    s32 compressed_data_len;
+    f.ReadArray(&compressed_data_len, 1);
+
+    auto compressed_data = std::make_unique<char[]>(compressed_data_len);
```
There needs to be a `if (compressed_data_len <= 0) break;` or similar here.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +629,99 @@ u64 GetUnixTimeOfSlot(int slot)
+
+    // We need to specify the output buffer's size for safety. This may exceed the positive bound
+    // of int, causing the buffer size to be interpreted as a negative value.
+    u32 max_decompress_size = (u32)std::min((u64)LZ4_MAX_INPUT_SIZE, size - total_bytes_read);
```
Please use `static_cast` instead of C-style casts.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +629,99 @@ u64 GetUnixTimeOfSlot(int slot)
+    u32 max_decompress_size = (u32)std::min((u64)LZ4_MAX_INPUT_SIZE, size - total_bytes_read);
+
+    int bytes_read =
+        LZ4_decompress_safe(compressed_data.get(), (char*)raw_buffer.data() + total_bytes_read,
```
`reinterpret_cast`

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -354,6 +306,71 @@ static std::map<double, int> GetSavedStates()
+    int compressed_buffer_size = LZ4_compressBound(bytes_to_compress);
+    auto compressed_buffer = std::make_unique<char[]>(compressed_buffer_size);
+    s32 compressed_len =
+        LZ4_compress_default((char*)raw_buffer + total_bytes_compressed, compressed_buffer.get(),
```
`reinterpret_cast`

malleoz replied:
This is going to be a bit ugly because raw_buffer is `const`. I assume it's fine if I do `reinterpret_cast<char*>(const_cast<u8*>(raw_buffer))`?

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -354,6 +306,71 @@ static std::map<double, int> GetSavedStates()
+    u64 bytes_left_to_compress = size - total_bytes_compressed;
+
+    int bytes_to_compress =
+        static_cast<int>(std::min((u64)LZ4_MAX_INPUT_SIZE, bytes_left_to_compress));
```
`static_cast` for the `(u64)`

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +629,99 @@ u64 GetUnixTimeOfSlot(int slot)
+      break;
+    }
+
+    total_bytes_read += bytes_read;
```
This needs a cast, you're adding a signed value to an unsigned one.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +745,50 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
-
-      i += new_len;
-    }
+    DecompressLZ4(buffer, extended_header.base_header.uncompressed_size, f);
```
`DecompressLZ4()` should return whether it succeeded or not, you don't want partially/incorrectly compressed data to be returned.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +745,50 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
-  if (strncmp(SConfig::GetInstance().GetGameID().c_str(), header.gameID, 6))
+  StateExtendedHeader extended_header;
+  f.ReadArray(&extended_header.base_header, 1);
+  // If StateExtendedHeader is ammended to include more than the base, add ReadBytes() calls here.
```
`amended`

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +745,50 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
+  StateExtendedHeader extended_header;
+  f.ReadArray(&extended_header.base_header, 1);
+  // If StateExtendedHeader is ammended to include more than the base, add ReadBytes() calls here.
+  // Do NOT use ReadArray, as this would prevent future revisions from reading in char arrays.
```
I still don't understand what this means.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +745,50 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
   }
-  else  // uncompressed
+  case (CompressionType::Uncompressed):
+  default:
```
The uncompressed read should only happen in the `CompressionType::Uncompressed` case (why are you putting those in parenthesis btw?). The default case should yell 'unknown compression type' and cancel.

malleoz replied:
The codebase at my job puts all their case statements in parentheses (I don't know why), so I accidentally had muscle memory here.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +745,50 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
-                                     std::string_view{header.gameID, std::size(header.gameID)}),
-                         2000);
-    return;
+    const std::string message =
+        fmt::format("Savestate extended header mismatch!\n\n"
+                    "Current: {}\n"
+                    "Savestate: {}",
+                    EXTENDED_HEADER_VERSION, extended_header.base_header.header_version);
+    Core::DisplayMessage(message, OSD::Duration::NORMAL);
```
This is a good log message but a poor user-facing string, for the user I'd just put 'state corrupted' or similar. Also I assume you want to return here?

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +629,99 @@ u64 GetUnixTimeOfSlot(int slot)
+    if (bytes_read < 0)
+    {
+      PanicAlertFmtT("Internal LZ4 Error - decompression failed ({0}, {1}, {2}) \n"
+                     "Try loading the state again",
```
I see that this line was here before but I'd argue it makes no sense? Why would loading the state a second time change anything?

malleoz replied:
I agree. I should've taken this out!

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +629,99 @@ u64 GetUnixTimeOfSlot(int slot)
+
+    loaded_str = "Dolphin " + oldest_version + " - " + newest_version;
+  }
+  else if (current_str != loaded_str || loaded_version != STATE_VERSION)
```
Err... I think this `current_str != loaded_str` breaks loading savestates from a different build even when the STATE_VERSION is the same, which should not be the case.

malleoz replied:
Oh my mistake, for some reason I thought it was intended behavior that we disallow savestates across Dolphin revisions, even if the state version is the same. Yeah I need to remove this check.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +745,50 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
+  default:
   {
-    const auto size = static_cast<size_t>(f.GetSize() - sizeof(StateHeader));
+    u32 header_len = sizeof(StateHeaderLegacy) + sizeof(StateHeaderVersion) +
```
This can overflow. Use `u64` instead.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +745,50 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
+    u32 header_len = sizeof(StateHeaderLegacy) + sizeof(StateHeaderVersion) +
+                     header.version_header.version_string_length + sizeof(StateExtendedBaseHeader) +
+                     extended_header.base_header.payload_offset;
+    const auto size = static_cast<size_t>(f.GetSize() - header_len);
```
And this can underflow, make sure `f.GetSize() >= header_len`.

AdmiralCurtiss reviewed (commented):
Lots of small things... This is looking pretty good in terms of general structure to me now though.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -378,48 +395,12 @@ static void CompressAndDumpState(CompressAndDumpState_args& save_args)
-  }
-  else  // uncompressed
-  {
+  if (s_use_compression)
+    CompressBufferToFile(buffer_data, buffer_size, f);
+  else
```
Sorta pre-existing but since you're here... if either the compression or the write fails, this should probably do something about that instead of blindly continuing the savestate write. We can do that in another PR since it's not new, if you want.

malleoz replied:
I'll do this in a new PR in the near future if that's alright!

malleoz commented:
Updated to address all suggestions that I marked as resolved! Thanks for the thorough review. I left a reply to clarify the remaining comments from @AdmiralCurtiss

malleoz commented:
Updated to remove the unnecessary comment I had left. I guess the only thing I'd like to clear up that remains is whether to keep the `buffer_data` pointer in `CompressAndDumpState()` as `const`. I think it makes sense to store it as `const` and just use `reinterpret_cast<char*>(const_cast<u8*>(raw_buffer))` but I don't know if this is best practice or not.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +752,56 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
       return;
     }
   }
+  default:
+    PanicAlertFmt("Unknown compression type {0}", extended_header.base_header.compression_type);
+  }
```
There should be a return in this case.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -354,6 +306,72 @@ static std::map<double, int> GetSavedStates()
+        static_cast<int>(std::min(static_cast<u64>(LZ4_MAX_INPUT_SIZE), bytes_left_to_compress));
+    int compressed_buffer_size = LZ4_compressBound(bytes_to_compress);
+    auto compressed_buffer = std::make_unique<char[]>(compressed_buffer_size);
+    s32 compressed_len = LZ4_compress_default(
+        reinterpret_cast<char*>(const_cast<u8*>(raw_buffer)) + total_bytes_compressed,
+        compressed_buffer.get(), bytes_to_compress, compressed_buffer_size);
```
You should be able to just do `reinterpret_cast<const char*>(raw_buffer) here, no?

malleoz replied:
Yes I can do that. I'm not sure why I missed that 😰

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +752,56 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
       PanicAlertFmt("Error reading bytes: {0}", size);
       return;
     }
   }
```
There's a `break;` missing here so it falls through into the default case.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -578,62 +752,56 @@ static void LoadFileStateData(const std::string& filename, std::vector<u8>& ret_
 
-  if (strncmp(SConfig::GetInstance().GetGameID().c_str(), header.gameID, 6))
+  StateExtendedHeader extended_header;
+  f.ReadArray(&extended_header.base_header, 1);
```
This needs its return value checked.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +631,104 @@ u64 GetUnixTimeOfSlot(int slot)
+  while (true)
+  {
+    s32 compressed_data_len;
+    f.ReadArray(&compressed_data_len, 1);
```
Return false on failure here.

AdmiralCurtiss commented on Source/Core/Core/State.cpp:
```diff
@@ -554,7 +631,104 @@ u64 GetUnixTimeOfSlot(int slot)
+    }
+
+    auto compressed_data = std::make_unique<char[]>(compressed_data_len);
+    f.ReadBytes(compressed_data.get(), compressed_data_len);
```
Return false on failure here.

malleoz commented:
Updated again! Thank you for the continued reviews. All suggestions should now be handled. Also I changed a few error messages from `Core::DisplayMessage` to `PanicAlertFmtT`. I believe anything pertaining to file read/parsing issues should use `PanicAlertFmtT` as this grabs the user's attention more to indicate savestate issues, although I'm not sure if this is the consensus.

AdmiralCurtiss reviewed (approved):
Seems fine now.

crashGG commented:
After many years, updating zstd to the latest upstream version will become attractive enough. According to official data, the encoding and decoding performance of the latest version of zstd has been objectively improved.
